Src/IronPython/Runtime/Operations/StringOps.cs

```diff
@@ -1701,7 +1706,7 @@ public static class StringOps {
 #if FEATURE_ENCODING
             // CLR's encoder exceptions have a 1-1 mapping w/ Python's encoder exceptions
             // so we just clone the encoding & set the fallback to throw in strict mode.
-            e = (Encoding)e.Clone();
+            if (e.IsReadOnly) e = (Encoding)e.Clone();
```

Comment:
I imagine you put this check for perf, however couldn't this cause all sorts of issues? For example, potentially mutating the .NET encoding that was given in `decode`? Race conditions with threading? Same comment for `DoEncode`.

Reply:
Yes, it was meant to reduce the amount of cloning happening, but it was a bad idea - as you pointed out, the encoding may have been already cloned by the user and would have been mutated here. As for race conditions, is `Clone` itself thread-safe?

Reply:
I have no idea is `Clone` itself is thread safe, it just does a memberwise copy. When I mentioned race conditions what I had in mind was more multiple `DoDecode` calls at the same time with an encoding that was already mutated:
1. (thread 1) e.DecoderFallback = bla
2. (thread 2) e.DecoderFallback = bla2
3. (thread 1) e.GetString() // this one ends up with the bla2 fallback
4. (thread 2) e.GetString()

Reply:
Yes, cloning will prevent such scenario. I think that the best rule here is to always (unconditionally) clone before setting a fallback and do it immediately before (to avoid unnecessary cloning if done earlier). I will change the relevant code in `Tokenizer` too to follow this rule, so that there is no assumption there on whence the encoding comes when returned by `TryGetEncoding`. 

But if `DoDecode` is meant to be thread-safe (and why not), I was thinking about such scenario:

1. (thread 1) e.Clone() // thread preempted half-way
2. (thread 2) e.SomeProperty = newValue
3. (thread 1) complete e.Clone() // object in an inconsistent state
4. (thread 1) e.GetString() // fails because of invalid state of e

If `MemberwiseCopy` copy is atomic then there is no problem, unless a derived encoding overrides `Clone`. And even if `MemberwiseCopy` is not atomic, it is no problem for the stock .NET encodings, as the only writable properties are the fallbacks, and they will be reset afterwards anyway. A derived encoding with a more complex cloning strategy will have to implement its own locks. Currently, there no such cases in IronPython but the codecs module is not finished yet. I am making a mental note to watch out for it.

Reply:
I agree unconditional cloning is the way to go.

I would argue that even if `Clone` is not atomic, then `DoDecode` would still be thread safe (by any sane definition of thread safety). If a user calls a method with an object and then spins up a thread to mutate the object, then it seems like they're asking for trouble. Anyway, other than being an interesting discussion I guess it's doesn't affect this PR. :smile:

Reply:
No, it was just to figure out what needs to be done in the future, if anything.

For this PR, I have the changes ready, just waiting for the tests to finish.